### PR TITLE
firefly-iii: Streamlined Derivation

### PR DESCRIPTION
**Pull request by savyajha**
## Description of changes

Removed building the `firefly-iii-assets` package and pulled everything into one package.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by savyajha** [comment]
@PatrickDaG I felt this refactoring of the package was needed because the original way of splitting the npm and composer builds was causing some of the built javascript files in the `firefly-iii-assets` package to not be transferred into the final `firefly-iii` package (from the `resources` directory, specifically).

This also allows `nix-update` to update all hashes in the derivation. Given that firefly-iii is [listed on repology](https://repology.org/project/firefly-iii/versions), I think we shouldn't need an explicit `updateScript`?

**Review by PatrickDaG — APPROVED** [review]
Looks great. Interestingly I've never noticed the missing files. I'm not sure if I was lucky or they're not needed but since the upstream packaged version also include them I'm in support of making sure they're added, to prevent unexpected breakage.

Something else I've noticed while comparing is that this package currently does not include translations, as they are downloaded by the GitHub release workflow upstream from Crowdin. I'm not sure if that would be possible for us or if there is anyway to get them, but I've also not really looked into it, so take this as just an observation.

**Comment by savyajha** [comment]
There's no way that I'm aware of of getting the Crowdin translations. Upstream uses a script for it which we can also use, but we'd need to be code owners for that. I tried looking at multiple ways of doing it and concluded the only way of getting the translations is to download the built package, remove the built stuff and then rebuild, which is excessive, imo.

File: pkgs/by-name/fi/firefly-iii/package.nix [hunk 1]
```diff
@@ -1,52 +1,84 @@
...
+    updateScript = nix-update-script { };
   };
 
   postInstall = ''
     mv $out/share/php/${pname}/* $out/
-    rm -R $out/share $out/storage $out/bootstrap/cache $out/public
-    cp -a ${assets} $out/public
+    rm -R $out/share $out/storage $out/bootstrap/cache
     ln -s ${dataDir}/storage $out/storage
     ln -s ${dataDir}/cache $out/bootstrap/cache
```
[the remark this answers is not in the document]
Comment (savyajha) [reply pkgs/by-name/fi/firefly-iii/package.nix:83]: Done!

**Comment by PatrickDaG** [comment]
> There's no way that I'm aware of of getting the Crowdin translations. Upstream uses a script for it which we can also use, but we'd need to be code owners for that. I tried looking at multiple ways of doing it and concluded the only way of getting the translations is to download the built package, remove the built stuff and then rebuild, which is excessive, imo.

Yeah. That's what I thought, too. Unfortunate, but I guess for now English has to suffice.

**Review by PatrickDaG — APPROVED** [review]

File: pkgs/by-name/fi/firefly-iii/package.nix [hunk 2]
```diff
@@ -1,52 +1,84 @@
...
   vendorHash = "sha256-EpMypgj6lZDz6T94bGoCUH9IVwh7VB4Ds08AcCsreRw=";
 
+  npmDeps = fetchNpmDeps {
+    inherit src;
+    name = "${pname}-npm-deps";
+    hash = npmDepsHash;
+  };
+
+  composerRepository = phpPackage.mkComposerRepository {
+    composer = phpPackage.packages.composer-local-repo-plugin;
```
Comment (drupol) [inline pkgs/by-name/fi/firefly-iii/package.nix:55]: Are you sure about requiring this? Normally it should be automatically provided, see: https://github.com/drupol/nixpkgs/blob/d083aa788eb7c24c121ce2a2f9ca58350eaa7235/pkgs/build-support/php/builders/v1/build-composer-repository.nix#L25-L26

Comment (savyajha) [reply pkgs/by-name/fi/firefly-iii/package.nix]: Fair enough. This is a remnant from back when I was trying to make this work. Changed!

**Review by PatrickDaG — APPROVED** [review]

File: pkgs/by-name/fi/firefly-iii/package.nix [hunk 3]
```diff
@@ -1,52 +1,83 @@
...
   inherit pname src version;
 
+  buildInputs = [ phpPackage ];
+
+  nativeBuildInputs = [
+    nodejs
+    nodejs.python
+    buildPackages.npmHooks.npmConfigHook
+    phpPackage.composerHooks.composerInstallHook
+    phpPackage.packages.composer-local-repo-plugin
```
Comment (drupol) [inline pkgs/by-name/fi/firefly-iii/package.nix:37]: Is this also needed?

Comment (savyajha) [reply pkgs/by-name/fi/firefly-iii/package.nix:37]: Yup
```bash
> Executing composerInstallBuildHook
> Setting COMPOSER_ROOT_VERSION to 6.1.18
> /nix/store/ly81bxpvvnpgg7x71zrdfw2pvbakyvq3-composer-install-hook.sh/nix-support/setup-hook: line 86: composer: command not found
```

**Review by drupol — APPROVED** [review]
LGTM

**Pull request merged**